File: app/src/main/java/com/example/bloodbank/ui/fragment/homeCycle/EditProfileFragment.java
package com.example.bloodbank.ui.fragment.homeCycle;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.widget.AppCompatSpinner;

import com.example.bloodbank.R;
import com.example.bloodbank.adapter.GeneralResponseAdapter;
import com.example.bloodbank.data.model.login.ClientData;
import com.example.bloodbank.data.model.login.Login;
import com.example.bloodbank.helper.DateModel;
import com.example.bloodbank.ui.activity.HomeActivity;
import com.example.bloodbank.ui.fragment.BaseFragment;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.example.bloodbank.data.api.ApiClient.getClient;
import static com.example.bloodbank.data.local.SharedPreferencesManger.API_TOKEN;
import static com.example.bloodbank.data.local.SharedPreferencesManger.LoadData;
import static com.example.bloodbank.data.local.SharedPreferencesManger.PASSWORD;
import static com.example.bloodbank.data.local.SharedPreferencesManger.SaveData;
import static com.example.bloodbank.data.local.SharedPreferencesManger.USER_DATA;
import static com.example.bloodbank.data.local.SharedPreferencesManger.loadUserData;
import static com.example.bloodbank.helper.HelperMethods.dismissProgressDialog;
import static com.example.bloodbank.helper.HelperMethods.getSpinnerCityData;
import static com.example.bloodbank.helper.HelperMethods.getSpinnerCityData2;
import static com.example.bloodbank.helper.HelperMethods.getSpinnerWithSelection;
import static com.example.bloodbank.helper.HelperMethods.showCalender;
import static com.example.bloodbank.helper.HelperMethods.showProgressDialog;
import static com.example.bloodbank.helper.HelperMethods.showToast;

public class EditProfileFragment extends BaseFragment {
    @BindView(R.id.fragment_edit_profile_ed_name)
    EditText fragmentEditProfileEdName;
    @BindView(R.id.fragment_edit_profile_ed_email)
    EditText fragmentEditProfileEdEmail;
    @BindView(R.id.fragment_edit_profile_tv_birth_date)
    TextView fragmentEditProfileTvBirthDate;
    @BindView(R.id.fragment_edit_profile_sp_blood_type)
    AppCompatSpinner fragmentEditProfileSpBloodType;
    @BindView(R.id.fragment_edit_profile_tv_donation_date)
    TextView fragmentEditProfileTvDonationDate;
    @BindView(R.id.fragment_edit_profile_sp_governorate)
    AppCompatSpinner fragmentEditProfileSpGovernorate;
    @BindView(R.id.fragment_edit_profile_sp_city)
    AppCompatSpinner fragmentEditProfileSpCity;
    @BindView(R.id.fragment_edit_profile_ed_phone)
    EditText fragmentEditProfileEdPhone;
    @BindView(R.id.fragment_edit_profile_ed_password)
    EditText fragmentEditProfileEdPassword;
    @BindView(R.id.fragment_edit_profile_ed_confirm_password)
    EditText fragmentEditProfileEdConfirmPassword;

    private static final String TAG = "EditProfileFragment";
    private String name, email, birth_date, phone, donation_last_date, password, password_confirmation;
    private int city_id, blood_type_id;
    private GeneralResponseAdapter governorateAdapter, cityAdapter, bloodTypeAdapter;
    private DateModel dateModel;

    private ClientData clientData;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        setupAcvtivity();
        View view = inflater.inflate(R.layout.fragment_edit_profile, container, false);
        ButterKnife.bind(this, view);

        dateModel = new DateModel("01", "01", "1970", "01-01-1970");

        clientData = loadUserData(getActivity());
        executSpinner();
        setData();
        return view;
    }

    private void executSpinner() {
        bloodTypeAdapter = new GeneralResponseAdapter(getActivity());
        governorateAdapter = new GeneralResponseAdapter(getActivity());
        cityAdapter = new GeneralResponseAdapter(getActivity());

        getSpinnerWithSelection(getClient().getbloodTypes(), bloodTypeAdapter, fragmentEditProfileSpBloodType, getString(R.string.blood_type),
                clientData.getClient().getBloodType().getId());

        getSpinnerCityData(getClient().getGovernorates(), governorateAdapter, fragmentEditProfileSpGovernorate, getString(R.string.governorate),
                clientData.getClient().getCity().getId(), new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
//                fragmentEditProfileSpCity.setSelection(clientData.getClient().getCity().getId());
                        getSpinnerWithSelection(getClient().getCities(position), cityAdapter, fragmentEditProfileSpCity, getString(R.string.city),
                                Integer.parseInt(clientData.getClient().getCity().getGovernorateId()));
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }
                });

        fragmentEditProfileSpGovernorate.setSelection(Integer.parseInt(clientData.getClient().getCity().getGovernorateId()));
        fragmentEditProfileSpCity.setSelection(clientData.getClient().getCity().getId());
    }

    private void setData() {
        fragmentEditProfileEdName.setText(clientData.getClient().getName());
        fragmentEditProfileEdEmail.setText(clientData.getClient().getEmail());
        fragmentEditProfileTvBirthDate.setText(clientData.getClient().getBirthDate());
        fragmentEditProfileTvDonationDate.setText(clientData.getClient().getDonationLastDate());
        fragmentEditProfileEdPhone.setText(clientData.getClient().getPhone());
        fragmentEditProfileEdPassword.setText(LoadData(getActivity(), PASSWORD));
        fragmentEditProfileEdConfirmPassword.setText(LoadData(getActivity(), PASSWORD));

    }

    private void editProfile() {
        name = fragmentEditProfileEdName.getText().toString();
        email = fragmentEditProfileEdEmail.getText().toString();
        birth_date = fragmentEditProfileTvBirthDate.getText().toString();
        phone = fragmentEditProfileEdPhone.getText().toString();
        donation_last_date = fragmentEditProfileTvDonationDate.getText().toString();
        password = fragmentEditProfileEdPassword.getText().toString();
        password_confirmation = fragmentEditProfileEdConfirmPassword.getText().toString();
        city_id = fragmentEditProfileSpCity.getSelectedItemPosition();
        blood_type_id = fragmentEditProfileSpBloodType.getSelectedItemPosition();

        showProgressDialog(getActivity(), getString(R.string.please_wait));
        getClient().signup(name, email, birth_date, city_id, phone, donation_last_date,
                password, password_confirmation, blood_type_id).enqueue(new Callback<Login>() {
            @Override
            public void onResponse(Call<Login> call, Response<Login> response) {
                dismissProgressDialog();
                try {
                    if (response.body().getStatus() == 1) {
                        showToast(getActivity(), response.body().getMsg());
                        SaveData(getActivity(), API_TOKEN, response.body().getData().getApiToken());
                        SaveData(getActivity(), USER_DATA, response.body().getData().getClient());
                        SaveData(getActivity(), PASSWORD, password);
                        startActivity(new Intent(getActivity(), HomeActivity.class));
                    } else {
                        showToast(getActivity(), response.body().getMsg());
                        Log.i(TAG, response.body().getMsg());
                    }
                } catch (Exception e) {
                    Log.i(TAG, response.body().getMsg());
                }
            }

            @Override
            public void onFailure(Call<Login> call, Throwable t) {

            }
        });
    }

    @OnClick({R.id.fragment_edit_profile_tv_birth_date, R.id.fragment_edit_profile_tv_donation_date, R.id.fragment_edit_profile_btn_edit})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.fragment_edit_profile_tv_birth_date:
                showCalender(getActivity(), getString(R.string.birth_date), fragmentEditProfileTvBirthDate, dateModel);
                break;
            case R.id.fragment_edit_profile_tv_donation_date:
                showCalender(getActivity(), getString(R.string.chose_donation_date), fragmentEditProfileTvDonationDate, dateModel);
                break;
            case R.id.fragment_edit_profile_btn_edit:
                editProfile();
                break;
        }
    }

    @Override
    public void onBack() {
        super.onBack();
    }
}

File: app/src/main/java/com/example/bloodbank/ui/fragment/homeCycle/EditProfileFragment.kt
package com.example.bloodbank.ui.fragment.homeCycle

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import com.example.bloodbank.R
import com.example.bloodbank.data.api.ApiClient.client
import com.example.bloodbank.data.local.SharedPreferencesManger.LoadData
import com.example.bloodbank.data.local.SharedPreferencesManger.SaveData
import com.example.bloodbank.data.local.SharedPreferencesManger.loadUserData
import com.example.bloodbank.data.model.login.ClientData
import com.example.bloodbank.data.model.login.Login
import com.example.bloodbank.databinding.FragmentEditProfileBinding
import com.example.bloodbank.extensions.inflateWithBinding
import com.example.bloodbank.helper.API_TOKEN
import com.example.bloodbank.helper.DateModel
import com.example.bloodbank.helper.HelperMethods.dismissProgressDialog
import com.example.bloodbank.helper.HelperMethods.getSpinnerCityData
import com.example.bloodbank.helper.HelperMethods.getSpinnerWithSelection
import com.example.bloodbank.helper.HelperMethods.showCalender
import com.example.bloodbank.helper.HelperMethods.showProgressDialog
import com.example.bloodbank.helper.HelperMethods.showToast
import com.example.bloodbank.helper.PASSWORD
import com.example.bloodbank.helper.USER_DATA
import com.example.bloodbank.ui.activity.HomeActivity
import com.example.bloodbank.ui.adapter.GeneralResponseAdapter
import com.example.bloodbank.ui.fragment.BaseFragment
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class EditProfileFragment : BaseFragment(), View.OnClickListener {

    private var name: String? = null
    private var email: String? = null
    private var birth_date: String? = null
    private var phone: String? = null
    private var donation_last_date: String? = null
    private var password_confirmation: String? = null
    private var city_id = 0
    private var blood_type_id = 0
    private var governorateAdapter: GeneralResponseAdapter? = null
    private var cityAdapter: GeneralResponseAdapter? = null
    private var bloodTypeAdapter: GeneralResponseAdapter? = null
    private var dateModel: DateModel? = null
    private var clientData: ClientData? = null
    private var password: String? = null
    private var loadPassword: String? = null

    lateinit var binding: FragmentEditProfileBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setupActivity()
        binding = inflater.inflateWithBinding(R.layout.fragment_edit_profile, container)
        dateModel = DateModel("01", "01", "1970", "01-01-1970")
        clientData = loadUserData(activity!!)
        loadPassword = LoadData(activity!!, PASSWORD)
        binding.clientData = clientData
        binding.password = loadPassword
        executSpinner()
        return binding.root
    }

    private fun executSpinner() {
        bloodTypeAdapter = GeneralResponseAdapter(activity)
        governorateAdapter = GeneralResponseAdapter(activity)
        cityAdapter = GeneralResponseAdapter(activity)
        getSpinnerWithSelection(client.getbloodTypes(), bloodTypeAdapter!!, binding.fragmentEditProfileSpBloodType, getString(R.string.blood_type),
                clientData!!.client!!.bloodType!!.id)
        getSpinnerCityData(client.governorates, governorateAdapter!!, binding.fragmentEditProfileSpGovernorate, getString(R.string.governorate),
                clientData!!.client!!.city!!.id, object : OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
//                fragmentEditProfileSpCity.setSelection(clientData.getClient().getCity().getId());
                getSpinnerWithSelection(client.getCities(position), cityAdapter!!, binding.fragmentEditProfileSpCity, getString(R.string.city), clientData!!.client!!.city!!.governorateId!!.toInt())
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        })
        binding.fragmentEditProfileSpGovernorate.setSelection(clientData!!.client!!.city!!.governorateId!!.toInt())
        binding.fragmentEditProfileSpCity.setSelection(clientData!!.client!!.city!!.id)
    }

    private fun editProfile() {
        name = binding.fragmentEditProfileEdName.text.toString()
        email = binding.fragmentEditProfileEdEmail.text.toString()
        birth_date = binding.fragmentEditProfileTvBirthDate.text.toString()
        phone = binding.fragmentEditProfileEdPhone.text.toString()
        donation_last_date = binding.fragmentEditProfileTvDonationDate.text.toString()
        password = binding.fragmentEditProfileEdPassword.text.toString()
        password_confirmation = binding.fragmentEditProfileEdConfirmPassword.text.toString()
        city_id = binding.fragmentEditProfileSpCity.selectedItemPosition
        blood_type_id = binding.fragmentEditProfileSpBloodType.selectedItemPosition
        showProgressDialog(activity, getString(R.string.please_wait))
        client.signup(name!!, email!!, birth_date!!, city_id, phone!!, donation_last_date!!,
                password!!, password_confirmation!!, blood_type_id).enqueue(object : Callback<Login> {
            override fun onResponse(call: Call<Login>, response: Response<Login>) {
                dismissProgressDialog()
                try {
                    if (response.body()!!.status == 1) {
                        showToast(activity, response.body()!!.msg)
                        SaveData(activity!!, API_TOKEN, response.body()!!.data!!.apiToken)
                        SaveData(activity!!, USER_DATA, response.body()!!.data!!.client)
                        SaveData(activity!!, PASSWORD, password)
                        startActivity(Intent(activity, HomeActivity::class.java))
                    } else {
                        showToast(activity, response.body()!!.msg)
                        Log.i(TAG, response.body()!!.msg.toString())
                    }
                } catch (e: Exception) {
                    Log.i(TAG, response.body()!!.msg.toString())
                }
            }

            override fun onFailure(call: Call<Login>, t: Throwable) {}
        })
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.fragment_edit_profile_tv_birth_date -> showCalender(activity, getString(R.string.birth_date), binding.fragmentEditProfileTvBirthDate, dateModel!!)
            R.id.fragment_edit_profile_tv_donation_date -> showCalender(activity, getString(R.string.chose_donation_date), binding.fragmentEditProfileTvDonationDate, dateModel!!)
            R.id.fragment_edit_profile_btn_edit -> editProfile()
        }
    }

    companion object {
        private const val TAG = "EditProfileFragment"
    }
}